lib/CL/devices/level0/level0-driver.hh

```diff
@@ -0,0 +1,308 @@
+  void copyRect(pocl_mem_identifier *DstMemId, cl_mem DstBuf,
+                pocl_mem_identifier *SrcMemId, cl_mem SrcBuf,
+                const size_t *__restrict__ const DstOrigin,
+                const size_t *__restrict__ const SrcOrigin,
+                const size_t *__restrict__ const Region,
+                size_t const DstRowPitch,
+                size_t const DstSlicePitch,
+                size_t const SrcRowPitch,
+                size_t const SrcSlicePitch);
+  void readRect(void *__restrict__ HostVoidPtr,
+                pocl_mem_identifier *SrcMemId, cl_mem SrcBuf,
+                const size_t *__restrict__ const BufferOrigin,
+                const size_t *__restrict__ const HostOrigin,
+                const size_t *__restrict__ const Region,
+                size_t const BufferRowPitch,
+                size_t const BufferSlicePitch,
+                size_t const HostRowPitch,
+                size_t const HostSlicePitch);
+  void writeRect(const void *__restrict__ HostVoidPtr,
+                 pocl_mem_identifier *DstMemId, cl_mem DstBuf,
+                 const size_t *__restrict__ const BufferOrigin,
+                 const size_t *__restrict__ const HostOrigin,
+                 const size_t *__restrict__ const Region,
+                 size_t const BufferRowPitch,
+                 size_t const BufferSlicePitch,
+                 size_t const HostRowPitch,
+                 size_t const HostSlicePitch);
+  void memFill(pocl_mem_identifier *DstMemId, cl_mem DstBuf,
+               size_t Size, size_t Offset,
+               const void *__restrict__ Pattern,
+               size_t PatternSize);
+  void mapMem(pocl_mem_identifier *SrcMemId, cl_mem SrcBuf,
+              mem_mapping_t *Map);
+  void unmapMem(pocl_mem_identifier *DstMemId, cl_mem DstBuf,
+                mem_mapping_t *map);
+
+  /* copies image to image, on the same device (or same global memory). */
+  void copyImageRect(cl_mem SrcImage, cl_mem DstImage,
+                     pocl_mem_identifier *SrcMemId,
+                     pocl_mem_identifier *DstMemId,
+                     const size_t *SrcOrigin,
+                     const size_t *DstOrigin,
+                     const size_t *Region);
+
+  /* copies a region from host OR device buffer to device image.
+   * clEnqueueCopyBufferToImage: src_mem_id = buffer,
+   *     src_host_ptr = NULL, src_row_pitch = src_slice_pitch = 0
+   * clEnqueueWriteImage: src_mem_id = NULL,
+   *     src_host_ptr = host pointer, src_offset = 0
+   */
+  void writeImageRect(cl_mem DstImage,
+                      pocl_mem_identifier *DstMemId,
+                      const void *__restrict__ src_HostPtr,
+                      pocl_mem_identifier *SrcMemId,
+                      const size_t *Origin, const size_t *Region,
+                      size_t SrcRowPitch, size_t SrcSlicePitch,
+                      size_t SrcOffset);
+
+  /* copies a region from device image to host or device buffer
```

Comment:
Check the "doxygen commenting guidelines" which are now in use (by Robin) and exemplified in poclu.

Reply:
these were actually just copied comments from device->ops callbacks in pocl_cl.h, removed them